File: zshref-web/src/lib/ranker/rank.ts
// WEB-MIRROR-OF: zshref-rs/src/nlp/rank.rs
//
// Pure ranker math. Inputs are pre-computed (query string, query vector,
// optional resolver hit) so this module has no embedder / corpus dependency.
//
// f32 math: every floating-point op funnels through Math.fround so the
// score sequence matches Rust's f32 evaluation bit-for-bit. Vector data
// flows through Float32Array so values read back as the exact f32 values
// Rust emitted. The parity-fixture test asserts byte-equality.

import type {
  Boosts,
  IndexedRecord,
  RankedMatch,
  RecordText,
  ResolverHit,
  SemanticScores,
  Tuning,
  VectorIndex
} from './types';
import type { Rules } from './rules';

const f = Math.fround;

function fAdd(...xs: number[]): number {
  // Left-fold matches Rust's `iter().sum::<f32>()` (fold from 0 with `+`).
  let s = 0;
  for (const x of xs) s = f(s + x);
  return s;
}

function fMul(a: number, b: number): number {
  return f(a * b);
}

function fSub(a: number, b: number): number {
  return f(a - b);
}

function fDiv(a: number, b: number): number {
  return f(a / b);
}

function fLn(x: number): number {
  return f(Math.log(x));
}

function clamp01(x: number): number {
  if (x < 0) return 0;
  if (x > 1) return 1;
  return x;
}

function dot(a: ArrayLike<number>, b: ArrayLike<number>): number {
  // Mirrors Rust's `iter().zip()` (stops at the shorter); in practice both are
  // DIMS-length, so the `?? 0` fallbacks never trigger (i stays in range).
  let s = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    s = f(s + f((a[i] ?? 0) * (b[i] ?? 0)));
  }
  return s;
}

export function rank(
  query: string,
  queryVec: Float32Array,
  resolverHit: ResolverHit | null,
  category: string | null,
  index: VectorIndex,
  rules: Rules
): RankedMatch[] {
  const q = query.toLowerCase();
  const penalties = categoryPenalties(index, rules);

  const out: RankedMatch[] = [];
  for (const rec of index.records) {
    if (category !== null && rec.text.category !== category) continue;
    const penalty = penalties.get(rec.text.category) ?? 0;
    out.push(scoreRecord(rec, queryVec, q, resolverHit, penalty, rules));
  }
  out.sort((a, b) => {
    if (a.score !== b.score) return b.score - a.score;
    if (a.rec.category !== b.rec.category) return a.rec.category < b.rec.category ? -1 : 1;
    if (a.rec.id !== b.rec.id) return a.rec.id < b.rec.id ? -1 : 1;
    return 0;
  });
  return out;
}

function scoreRecord(
  rec: IndexedRecord,
  queryVec: Float32Array,
  q: string,
  resolverHit: ResolverHit | null,
  categoryPenalty: number,
  rules: Rules
): RankedMatch {
  const semantic: SemanticScores = {
    structured: dot(queryVec, rec.vectors.structured),
    body: dot(queryVec, rec.vectors.body),
    expanded: dot(queryVec, rec.vectors.expanded)
  };
  const b = computeBoosts(rec.text, q, resolverHit, rules);

  const bodyWords = countWords(rec.text.body);
  const [bodyW, structW, expW] = semanticWeights(bodyWords, rules.tuning.semantic_weights);

  const semanticScore = fAdd(
    fMul(bodyW, semantic.body),
    fMul(structW, semantic.structured),
    fMul(expW, semantic.expanded)
  );

  const score = fSub(fAdd(semanticScore, b.category, b.resolver, b.lexical), categoryPenalty);

  return { rec: rec.text, score, debug: { semantic, boosts: b } };
}

function countWords(s: string): number {
  // Mirrors Rust's `str::split_whitespace().count()`.
  return s.split(/\s+/).filter((w) => w.length > 0).length;
}

// Mirror of rank.rs `semantic_weights`.
function semanticWeights(
  bodyWords: number,
  sw: Tuning['semantic_weights']
): [number, number, number] {
  const body = f(sw.body);
  const structured = f(sw.structured);
  const expanded = fSub(fSub(1, body), structured);
  const ramp = Math.max(fSub(1, fDiv(f(bodyWords), f(sw.short_body.length_scale))), 0);
  const shift = Math.min(Math.max(fMul(f(sw.short_body.strength), ramp), 0), body);
  return [fSub(body, shift), structured, fAdd(expanded, shift)];
}

function categoryPenalty(recordCount: number, maxRecords: number, rules: Rules): number {
  if (maxRecords <= 1) return 0;
  const rarity = f(1 - f(fLn(Math.max(recordCount, 1)) / fLn(maxRecords)));
  return fMul(f(rules.tuning.penalties.category_rarity_max), clamp01(rarity));
}

function categoryPenalties(index: VectorIndex, rules: Rules): Map<string, number> {
  const counts = new Map<string, number>();
  for (const rec of index.records) {
    counts.set(rec.text.category, (counts.get(rec.text.category) ?? 0) + 1);
  }
  let maxRecords = 0;
  for (const c of counts.values()) if (c > maxRecords) maxRecords = c;
  const out = new Map<string, number>();
  for (const [cat, count] of counts) {
    out.set(cat, categoryPenalty(count, maxRecords, rules));
  }
  return out;
}

// Mirror of rank.rs `BoostWeights::{exact_word, resolver}`.
function exactWordBoost(b: Tuning['boosts']): number {
  return f(f(b.category) + f(b.exact_word_increment));
}

function resolverBoost(b: Tuning['boosts']): number {
  return f(exactWordBoost(b) + f(b.resolver_increment));
}

function computeBoosts(
  rec: RecordText,
  q: string,
  resolverHit: ResolverHit | null,
  rules: Rules
): Boosts {
  const b = rules.tuning.boosts;
  const categoryWord = rec.category.toLowerCase();
  const labelWord = rec.category_label.toLowerCase();
  const category = q.includes(categoryWord) || q.includes(labelWord) ? f(b.category) : 0;
  const resolver =
    resolverHit && resolverHit.category === rec.category && resolverHit.id === rec.id
      ? resolverBoost(b)
      : 0;
  const id = rec.id.toLowerCase();
  const display = rec.display.toLowerCase();
  const wordExact = significantWords(q, rules)
    .filter((w) => isDiscriminating(w, rules))
    .some((w) => {
      const lw = w.toLowerCase();
      return lw === id || lw === display;
    });
  // Symbolic surface match: zsh users name operators and special parameters by
  // their literal symbol ("$?", ">>", "<<<"), which is punctuation, so
  // significantWords drops it. Match those tokens against the record's id and
  // the symbolic head of its display — the punctuation analogue of wordExact.
  const symbolExact = symbolTokens(q).some((t) => t === id || symbolHead(display) === t);
  const exactWord = wordExact || symbolExact ? exactWordBoost(b) : 0;
  const overlap = f(wordOverlap(rec, q, rules));
  const wo = b.word_overlap;
  // Mirror of rank.rs `overlap_boost`.
  const overlapBoost = fDiv(fMul(f(wo.scale), overlap), fAdd(overlap, f(wo.half_sat)));
  const lexical = f(exactWord + overlapBoost);
  return {
    category,
    resolver,
    lexical
  };
}

function isDiscriminating(word: string, rules: Rules): boolean {
  if (word.length < rules.tuning.lexical.min_discriminating_word_len) return false;
  const lower = word.toLowerCase();
  const sw = rules.stopwords;
  if (sw.generic.some((s) => s === lower)) return false;
  if (sw.discriminating_extra.some((s) => s === lower)) return false;
  return true;
}

function wordOverlap(rec: RecordText, q: string, rules: Rules): number {
  const recText = `${rec.id} ${rec.display} ${rec.structured} ${rec.body} ${rec.expanded}`.toLowerCase();
  let count = 0;
  for (const w of significantWords(q, rules)) {
    if (recText.includes(w)) count++;
  }
  return count;
}

// Mirror of rank.rs `symbol_tokens`.
// Literal symbol tokens in `q`: whitespace tokens that bear punctuation or are
// `$`-sigiled parameter refs, lowercased, with surrounding quotes and one
// leading `$` stripped ("$?" -> "?"). Exactly what significantWords discards,
// yet how zsh names its operators and special parameters.
function symbolTokens(q: string): string[] {
  const out: string[] = [];
  for (const t of q.split(/\s+/)) {
    if (t.length === 0) continue;
    if (!t.startsWith('$') && !/[^A-Za-z0-9]/.test(t)) continue;
    const trimmed = t.replace(/^['"`]+/, '').replace(/['"`]+$/, '');
    const stripped = (trimmed.startsWith('$') ? trimmed.slice(1) : trimmed).toLowerCase();
    if (stripped.length > 0) out.push(stripped);
  }
  return out;
}

// Mirror of rank.rs `symbol_head`.
// Leading run of operator characters in a display form — the symbol before any
// alphanumeric operand placeholder: ">> word" -> ">>", "?" -> "?",
// "auto_cd" -> null. Lets a bare-operator query match a sig-shaped record.
function symbolHead(display: string): string | null {
  const m = display.search(/[A-Za-z0-9 ]/);
  const end = m === -1 ? display.length : m;
  return end > 0 ? display.slice(0, end) : null;
}

function significantWords(s: string, rules: Rules): string[] {
  const min = rules.tuning.lexical.min_significant_word_len;
  const sw = rules.stopwords;
  const out: string[] = [];
  for (const word of s.split(/[^A-Za-z0-9]+/)) {
    if (word.length < min) continue;
    if (sw.generic.some((stop) => stop === word)) continue;
    out.push(word);
  }
  return out;
}

File: zshref-rs/src/nlp/rank.rs
// WEB-MIRRORED-IN: zshref-web/src/lib/ranker/rank.ts
//
// Pure ranker math — all rank-time (no embedder / corpus dependency).
// Inputs are pre-computed (query string, query vector, optional resolver
// hit) so the TS mirror in zshref-web can re-implement it 1:1.

use crate::nlp::index::{IndexedRecord, VectorIndex};
use crate::nlp::retrieval_text::RecordText;
use crate::nlp::rules::{stopwords, BoostWeights, SemanticWeights, Tuning};
use serde::Serialize;
use std::collections::HashMap;

/// Resolver-resolved (category, id) pair from outside the ranker.
pub type ResolverHit = (String, String);

#[derive(Debug, Serialize)]
pub struct RankedMatch<'a> {
    pub rec: &'a RecordText,
    pub score: f32,
    pub debug: RankDebug,
}

#[derive(Clone, Debug, Serialize)]
pub struct RankDebug {
    pub semantic: SemanticScores,
    pub boosts: Boosts,
}

#[derive(Clone, Debug, Serialize)]
pub struct SemanticScores {
    pub structured: f32,
    pub body: f32,
    pub expanded: f32,
}

#[derive(Clone, Debug, Serialize)]
pub struct Boosts {
    pub category: f32,
    pub resolver: f32,
    pub lexical: f32,
}

pub fn rank<'a>(
    query: &str,
    query_vec: &[f32],
    resolver_hit: Option<&ResolverHit>,
    category: Option<&str>,
    index: &'a VectorIndex,
    tuning: &Tuning,
) -> Vec<RankedMatch<'a>> {
    let q = query.to_ascii_lowercase();
    let penalties = category_penalties(index, tuning);

    let mut out: Vec<RankedMatch<'a>> = index
        .records
        .iter()
        .filter(|rec| category.is_none_or(|cat| rec.text.category == cat))
        .map(|rec| {
            let penalty = penalties
                .get(rec.text.category.as_str())
                .copied()
                .unwrap_or(0.0);
            score_record(rec, query_vec, &q, resolver_hit, penalty, tuning)
        })
        .collect();
    out.sort_by(|a, b| {
        b.score
            .total_cmp(&a.score)
            .then_with(|| a.rec.category.cmp(&b.rec.category))
            .then_with(|| a.rec.id.cmp(&b.rec.id))
    });
    out
}

fn score_record<'a>(
    rec: &'a IndexedRecord,
    query_vec: &[f32],
    q: &str,
    resolver_hit: Option<&ResolverHit>,
    category_penalty: f32,
    tuning: &Tuning,
) -> RankedMatch<'a> {
    let semantic = SemanticScores {
        structured: dot(query_vec, &rec.vectors.structured),
        body: dot(query_vec, &rec.vectors.body),
        expanded: dot(query_vec, &rec.vectors.expanded),
    };
    let boosts = boosts(&rec.text, q, resolver_hit, tuning);

    let body_words = rec.text.body.split_whitespace().count();
    let (body_w, struct_w, exp_w) = semantic_weights(body_words, &tuning.semantic_weights);
    let semantic_score =
        body_w * semantic.body + struct_w * semantic.structured + exp_w * semantic.expanded;

    RankedMatch {
        rec: &rec.text,
        score: semantic_score + boosts.category + boosts.resolver + boosts.lexical
            - category_penalty,
        debug: RankDebug { semantic, boosts },
    }
}

/// Effective (body, structured, expanded) weights. `expanded` is derived as
/// `1 − body − structured` (on the simplex by construction), then mass is
/// shifted body → expanded the shorter the body. Continuous — no threshold cliff.
fn semantic_weights(body_words: usize, sw: &SemanticWeights) -> (f32, f32, f32) {
    let body = sw.body;
    let structured = sw.structured;
    let expanded = 1.0 - body - structured;
    let ramp = (1.0 - body_words as f32 / sw.short_body.length_scale).max(0.0);
    let shift = (sw.short_body.strength * ramp).clamp(0.0, body);
    (body - shift, structured, expanded + shift)
}

/// Data-derived unusualness penalty: smaller categories need slightly higher
/// semantic score to outrank broad categories.
fn category_penalty(record_count: usize, max_records: usize, tuning: &Tuning) -> f32 {
    if max_records <= 1 {
        return 0.0;
    }
    let rarity = 1.0 - (record_count.max(1) as f32).ln() / (max_records as f32).ln();
    tuning.penalties.category_rarity_max * rarity.clamp(0.0, 1.0)
}

fn category_penalties<'a>(index: &'a VectorIndex, tuning: &Tuning) -> HashMap<&'a str, f32> {
    let mut counts: HashMap<&str, usize> = HashMap::new();
    for rec in &index.records {
        *counts.entry(rec.text.category.as_str()).or_insert(0) += 1;
    }
    let max_records = counts.values().copied().max().unwrap_or(0);
    counts
        .into_iter()
        .map(|(cat, count)| (cat, category_penalty(count, max_records, tuning)))
        .collect()
}

fn boosts(
    rec: &RecordText,
    q: &str,
    resolver_hit: Option<&ResolverHit>,
    tuning: &Tuning,
) -> Boosts {
    let b = &tuning.boosts;
    let category = if q.contains(&rec.category.to_ascii_lowercase())
        || q.contains(&rec.category_label.to_ascii_lowercase())
    {
        b.category
    } else {
        0.0
    };
    let resolver = resolver_hit
        .filter(|(cat, id)| cat == &rec.category && id == &rec.id)
        .map(|_| b.resolver())
        .unwrap_or(0.0);
    let id = rec.id.to_ascii_lowercase();
    let display = rec.display.to_ascii_lowercase();
    let word_exact = significant_words(q, tuning)
        .into_iter()
        .filter(|word| is_discriminating_word(word, tuning))
        .any(|word| {
            let w = word.to_ascii_lowercase();
            w == id || w == display
        });
    // Symbolic surface match: zsh users name operators and special parameters by
    // their literal symbol ("$?", ">>", "<<<"), which is punctuation, so
    // significant_words drops it. Match those tokens against the record's id and
    // the symbolic head of its display — the punctuation analogue of word_exact.
    let symbol_exact = symbol_tokens(q)
        .iter()
        .any(|t| *t == id || symbol_head(&display).is_some_and(|h| h == t));
    let exact_word = if word_exact || symbol_exact {
        b.exact_word()
    } else {
        0.0
    };
    let lexical = exact_word + overlap_boost(word_overlap(rec, q, tuning), b);

    Boosts {
        category,
        resolver,
        lexical,
    }
}

/// Smooth saturating lexical-overlap boost over the overlap count `n`:
/// `scale · n / (n + half_sat)` — monotone, asymptote `scale`, half at `half_sat`.
fn overlap_boost(n: usize, b: &BoostWeights) -> f32 {
    let n = n as f32;
    let wo = &b.word_overlap;
    wo.scale * n / (n + wo.half_sat)
}

fn is_discriminating_word(word: &str, tuning: &Tuning) -> bool {
    if word.len() < tuning.lexical.min_discriminating_word_len {
        return false;
    }
    let sw = stopwords();
    let lower = word.to_ascii_lowercase();
    !sw.generic.iter().any(|s| s == &lower) && !sw.discriminating_extra.iter().any(|s| s == &lower)
}

fn word_overlap(rec: &RecordText, q: &str, tuning: &Tuning) -> usize {
    let rec_text = format!(
        "{} {} {} {} {}",
        rec.id, rec.display, rec.structured, rec.body, rec.expanded
    )
    .to_ascii_lowercase();
    significant_words(q, tuning)
        .into_iter()
        .filter(|word| rec_text.contains(word))
        .count()
}

fn significant_words<'a>(s: &'a str, tuning: &Tuning) -> Vec<&'a str> {
    let sw = stopwords();
    let min = tuning.lexical.min_significant_word_len;
    s.split(|c: char| !c.is_ascii_alphanumeric())
        .filter(move |word| word.len() >= min)
        .filter(|word| !sw.generic.iter().any(|stop| stop == *word))
        .collect()
}

/// Literal symbol tokens in `q`: whitespace tokens that bear punctuation or are
/// `$`-sigiled parameter refs, lowercased, with surrounding quotes and one
/// leading `$` stripped ("$?" -> "?"). These are exactly what `significant_words`
/// discards, yet they are how zsh names its operators and special parameters.
fn symbol_tokens(q: &str) -> Vec<String> {
    q.split_whitespace()
        .filter(|t| t.starts_with('$') || t.chars().any(|c| !c.is_ascii_alphanumeric()))
        .map(|t| t.trim_matches(|c| matches!(c, '\'' | '"' | '`')))
        .map(|t| t.strip_prefix('$').unwrap_or(t).to_ascii_lowercase())
        .filter(|t| !t.is_empty())
        .collect()
}

/// Leading run of operator characters in a display form — the symbol before any
/// alphanumeric operand placeholder: ">> word" -> Some(">>"), "?" -> Some("?"),
/// "auto_cd" -> None. Lets a bare-operator query match a sig-shaped record.
fn symbol_head(display: &str) -> Option<&str> {
    let end = display
        .find(|c: char| c.is_ascii_alphanumeric() || c == ' ')
        .unwrap_or(display.len());
    (end > 0).then_some(&display[..end])
}

fn dot(a: &[f32], b: &[f32]) -> f32 {
    a.iter().zip(b).map(|(x, y)| x * y).sum()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::nlp::rules::tuning;

    #[test]
    fn exact_id_match_gives_lexical_boost() {
        let rec = RecordText {
            category: "option".into(),
            category_label: "option".into(),
            id: "autocd".into(),
            display: "AUTO_CD".into(),
            sub_kind: None,
            title: String::new(),
            md_body: String::new(),
            structured: String::new(),
            body: String::new(),
            expanded: String::new(),
        };
        let exact = boosts(&rec, "autocd", None, tuning());
        let query_with_category = boosts(&rec, "option autocd please", None, tuning());
        // Both queries contain the discriminating word "autocd" and get the
        // exact-word boost, so lexical scores can be equal.
        assert!(exact.lexical > 0.0);
        assert!(query_with_category.category > 0.0);
    }

    #[test]
    fn symbol_tokens_strip_sigil_and_keep_operators() {
        assert_eq!(symbol_tokens("the $? param"), vec!["?"]);
        assert_eq!(symbol_tokens("redirection >>"), vec![">>"]);
        assert_eq!(symbol_tokens("$0"), vec!["0"]); // sigiled even if alnum after strip
        assert!(symbol_tokens("list all background jobs").is_empty());
    }

    #[test]
    fn symbol_head_is_the_operator_prefix() {
        assert_eq!(symbol_head(">> word"), Some(">>"));
        assert_eq!(symbol_head("?"), Some("?"));
        assert_eq!(symbol_head("auto_cd"), None); // leading alnum -> no symbol head
    }

    #[test]
    fn symbol_query_matches_param_and_operator_records() {
        let param = RecordText {
            category: "special_param".into(),
            category_label: "special parameter".into(),
            id: "?".into(),
            display: "?".into(),
            sub_kind: None,
            title: String::new(),
            md_body: String::new(),
            structured: String::new(),
            body: String::new(),
            expanded: String::new(),
        };
        let redir = RecordText {
            category: "redirection".into(),
            id: ">>_word".into(),
            display: ">> word".into(),
            ..param.clone()
        };
        // "$?" names the `?` param; ">>" names `>>_word` via its display's
        // symbolic head — both fire the lexical boost. A prose word does not.
        assert!(boosts(&param, "the $? param", None, tuning()).lexical > 0.0);
        assert!(boosts(&redir, "redirection >>", None, tuning()).lexical > 0.0);
        assert_eq!(
            boosts(&param, "list background jobs", None, tuning()).lexical,
            0.0
        );
    }

    use crate::nlp::rules::ShortBodyDeform;

    fn sw(body: f32, structured: f32, strength: f32) -> SemanticWeights {
        SemanticWeights {
            body,
            structured,
            short_body: ShortBodyDeform {
                strength,
                length_scale: 10.0,
            },
        }
    }

    #[test]
    fn semantic_weights_derive_expanded() {
        // expanded = 1 − body − structured; the triple sums to 1 by construction.
        let (b, s, e) = semantic_weights(1000, &sw(0.70, 0.20, 0.0));
        assert!((b - 0.70).abs() < 1e-6);
        assert!((s - 0.20).abs() < 1e-6);
        assert!((e - 0.10).abs() < 1e-6);
        assert!((b + s + e - 1.0).abs() < 1e-6);
    }

    #[test]
    fn short_body_shift_is_continuous_with_exact_endpoints() {
        let w = sw(0.70, 0.20, 0.10);
        // L ≥ length_scale: base mix, no shift.
        assert!((semantic_weights(10, &w).0 - 0.70).abs() < 1e-6);
        // L = 0: full strength shift body → expanded (the old short-body triple).
        let z = semantic_weights(0, &w);
        assert!((z.0 - 0.60).abs() < 1e-6);
        assert!((z.1 - 0.20).abs() < 1e-6);
        assert!((z.2 - 0.20).abs() < 1e-6);
        // Strictly monotone across the ramp — no discontinuity.
        let a = semantic_weights(2, &w).0;
        let b = semantic_weights(7, &w).0;
        assert!(0.60 < a && a < b && b < 0.70);
    }

    #[test]
    fn boosts_are_reliability_ordered() {
        let b = &tuning().boosts;
        assert!(b.category <= b.exact_word());
        assert!(b.exact_word() <= b.resolver());
    }

    #[test]
    fn overlap_boost_saturates_monotonically() {
        let b = &tuning().boosts;
        assert_eq!(overlap_boost(0, b), 0.0);
        let one = overlap_boost(1, b);
        let many = overlap_boost(100, b);
        assert!(one > 0.0 && one < many);
        // Smooth saturation never reaches the asymptote.
        assert!(many < b.word_overlap.scale);
    }

    #[test]
    fn prose_overlap_beats_broad_name_containment() {
        let aliases = RecordText {
            category: "option".into(),
            category_label: "option".into(),
            id: "aliases".into(),
            display: "ALIASES".into(),
            sub_kind: None,
            title: String::new(),
            md_body: String::new(),
            structured: "id: aliases".into(),
            body: "Expand aliases.".into(),
            expanded: String::new(),
        };
        let complete = RecordText {
            id: "completealiases".into(),
            display: "COMPLETE_ALIASES".into(),
            body: "Prevents aliases before completion is attempted.".into(),
            structured: "id: completealiases".into(),
            expanded: String::new(),
            ..aliases.clone()
        };
        // Query has more discriminating words in complete's body/structured
        // than in aliases's — overlap should favour complete.
        let q = "prevents expansion before completion";
        assert!(
            boosts(&complete, q, None, tuning()).lexical
                > boosts(&aliases, q, None, tuning()).lexical
        );
    }
}
